Add better ExUnit formatting when missing head of list

devonestes opened the pull request:
The current implementation of the Myers algorithm for diffing didn't
give very helpful diff information when comparing lists in the ExUnit
formatter. This is a very naieve implementation, and at the moment it
is only for lists, but I can see that we might conceivably want to have
this "missing head" diff show up for strings as well.

For making just this work, I sort of skipped the whole Myers algorithm
and just hand built the diff script since it follows a pretty simple
pattern. If we want to expand this further, though, we might need to
implement the reverse of the Myers algorithm.

Resolves #6768

devonestes commented:
Got it - I'll go with the full algorithm them.

devonestes commented:
So I've expanded this a bit, right now just for non-keyword lists. Here are some examples of the current output:

![first](https://user-images.githubusercontent.com/8422484/39869187-ec79a482-545c-11e8-896f-4a2ed262ac37.png)
![second](https://user-images.githubusercontent.com/8422484/39869189-ecace13a-545c-11e8-9b2d-8cbcc695b6a3.png)
![third](https://user-images.githubusercontent.com/8422484/39869190-ecc2c680-545c-11e8-8ece-64ae83ce8671.png)
![fourth](https://user-images.githubusercontent.com/8422484/39869188-ec97b54e-545c-11e8-953d-1aa7dc280b35.png)

The first comma in the third and fourth example sort of bothers me, but changing that would require changing existing behavior found [here](https://github.com/elixir-lang/elixir/blob/a2b27f4d0212dd7d92166d332df7760dcd85a51f/lib/ex_unit/lib/ex_unit/diff.ex#L186), so I didn't want to change that for now unless y'all want it changed to better support this. I mean, technically this script is correct, but it would make more sense to me personally if that first comma was a `:del` or `:ins` instead of an `:eq`. I can see how some people might prefer this way, though, so I didn't want to change it just yet.

I realized that when we pull the equal elements from the front and back of the list, the only time that we do anything other than stick with what we're doing now is when the list of common elements at the head of the list was empty and the list of elements at the back of the list wasn't empty, so I substituted that with checking the head and tail of the list to ensure that the tails were equal and the heads weren't equal. In all other cases we would continue with the current behavior, so I left that alone.

If y'all think this is on the right track I'll continue implementing this for the other data types!

josevalim commented on lib/ex_unit/lib/ex_unit/diff.ex:
```diff
@@ -269,11 +314,9 @@ defmodule ExUnit.Diff do
+  defp script_keyword_inner(_pair1, _pair2, true), do: nil
 
-  defp script_keyword_inner(elem1, elem2, false),
-    do: script(elem1, elem2)
```
You are not running the latest version of the formatter, that's why this change happened. Make sure you have 1.6.5 in your system or run `bin/elixir bin/mix format`.

devonestes replied:
oh, yeah, sorry! My vim integration isn't set for `bin/elixir bin/mix` 😢

josevalim commented:
Hi @devonestes! I looked at the current implementation and it feels is more complex than it has to be. Maybe that's the source of the comma issue, I am not sure, but here is what I would try:

```elixir
case diff_subset(list1, list2) do
  {:ok, diff} -> diff
  :error -> ...invokethecodefromtoday...
end
```

where `diff_subset(list1, list2)` would be implemented like this:

```elixir
def diff_subset(list1, list2) do
  {list1, list2, prefix} = take_common_path(list1, list2, [])
  {list1, list2, suffix} = take_common_path(Enum.reverse(list1), Enum.reverse(list2), [])

  if list1 == [] or list2 == [] do
    {:ok, wrap_in(:eq, Enum.reverse(prefix)) ++ wrap_in(:ins, Enum.reverse(list1)) ++ wrap_in(:del, Enum.reverse(list2)) ++ wrap_in(:eq, suffix)}
  else
    :error
  end
end

defp take_common_path([h | t1], [h | t2], acc), do: take_common_path(t1, t2, [h | acc])
defp take_common_path(t1, t2, acc), do: {t1, t2, acc}

defp wrap_in(tag, items), do: Enum.map(items, &{tag, &1})
```

Roughly pseudo code, probably has errors around the corners, but it should point you towards a simpler solution. It may also be that I am missing some context from the code that makes it more complex. In any case, the above is enough for the "take both prefix and suffix out, assert one of the remaining ones is empty and then generate a diff".

josevalim commented:
Regarding the other types, the only remaining data types where this would apply to are tuples (which probably share the logic with lists) and strings. But let's make sure we got all list cases covered first.

josevalim commented on lib/ex_unit/lib/ex_unit/diff.ex:
```diff
@@ -143,22 +143,67 @@ defmodule ExUnit.Diff do
+    [{:eq, "["}, result, {:eq, "]"}]
   end
 
+  defp format_each_fragment([{:diff, script}], [], _keywords?), do: script
```
This diff is also because of the formatter.

josevalim commented on lib/ex_unit/test/ex_unit/diff_test.exs:
```diff
@@ -125,6 +125,39 @@ defmodule ExUnit.DiffTest do
     assert script(list1, list2) == expected
 
     assert script([], []) == [eq: "[]"]
+
```
Let's create a new test case especifically about testing lists that are subsets instead of adding more examples to those already large cases :)

josevalim commented:
@devonestes can you please rebase once again before moving forward? I have noticed that tuples and lists could share their logic, so I have refactored the code. This means that once you are done with lists, you will be done with tuples too. :D

devonestes commented:
Ok, rebased! For some reason it didn't click for me until I saw you implementation that we were looking for subsets! I was still doing something totally different.

I think this is along the lines of what y'all want now, and I also added tests for keyword lists and tuples since those are now covered as well. I didn't think we needed additional tests for charlists since those are just lists without any special behavior in the diffing.

devonestes commented:
Oh, and it looks like strings already are looking for substrings and formatting appropriately, which means that all data types we want to support should be taken care of. I tried writing a failing test and couldn't make it fail, so that's a good thing.

Also, maybe this is definitely for a separate issue, but there seems to be a pretty significant amount of duplication between the implementation of the Myers algorithm in `ExUnit.Diff` and the implementation in `List.myers_difference`. Is there a reason we can't or shouldn't unify those two?

josevalim commented:
@devonestes I have added the word "non-empty" to the test cases since we have subsets and supersets in other tests, it just happens that they are empty or equal lists. What do you think?

josevalim commented:
@devonestes can you also please add subset tests for tuples? thank you!

lexmag commented:
Thank you @devonestes. 💛
